src/spatialmath/quaternion.rs

```diff
@@ -0,0 +1,371 @@
+/// A Rust implementation of Quaternion, we use this instead of existing packages
+/// because we want a C-safe representational structure. These quaternions use
+/// the Real-I-J-K standard, so those using JPL or other standards should take
+/// care to convert before initializing a Quaternion struct from the library
+#[repr(C)]
+#[derive(Copy, Clone, Debug)]
+pub struct Quaternion {
+    pub real: f64,
+    pub i: f64,
+    pub j: f64,
+    pub k: f64,
+}
+
+impl Quaternion {
+    /// Initializes a quaternion from a real component and an
+    /// imaginary 3-vector
+    pub fn new_from_vector(real: f64, imag: Vector3) -> Self {
+        Self{real: real, i: imag.x, j: imag.y, k: imag.z}
+    }
+
+    /// Initializes a Quaternion from its raw components
+    pub fn new(real: f64, i: f64, j: f64, k: f64) -> Self {
+        Self{real, i, j, k}
+    }
+
+    /// Converts from euler angles to a quaternion. The euler angles are expected to
+    /// be represented according to the Tait-Bryan formalism and applied in the Z-Y'-X"
+    /// order (where Z -> yaw, Y -> pitch, X -> roll)
+    pub fn from_euler_angles(roll: f64, pitch: f64, yaw: f64) -> Self {
+        let roll_cos = (roll * 0.5).cos();
+        let roll_sin = (roll * 0.5).sin();
+
+        let pitch_cos = (pitch * 0.5).cos();
+        let pitch_sin = (pitch * 0.5).sin();
+
+        let yaw_cos = (yaw * 0.5).cos();
+        let yaw_sin = (yaw * 0.5).sin();
+
+        let real = (roll_cos * pitch_cos * yaw_cos) + (roll_sin * pitch_sin * yaw_sin);
+        let i = (roll_sin * pitch_cos * yaw_cos) - (roll_cos * pitch_sin * yaw_sin);
+        let j = (roll_cos * pitch_sin * yaw_cos) + (roll_sin * pitch_cos * yaw_sin);
+        let k = (roll_cos * pitch_cos * yaw_sin) - (roll_sin * pitch_sin * yaw_cos);
+
+        Quaternion::new(real, i, j, k)
+    }
+
+    /// Converts a quaternion into euler angles. The euler angles are 
+    /// represented according to the Tait-Bryan formalism and applied 
+    /// in the Z-Y'-X" order (where Z -> yaw, Y -> pitch, X -> roll). 
+    /// The return value is a list of [roll, pitch, yaw]
+    pub fn to_euler_angles(&self) -> [f64;3] {
+        // get a normalized version of the quaternion
+        let quat = self.get_normalized();
+
+        // calculate yaw
+        let yaw_sin_pitch_cos = 2.0 * ((quat.real * quat.k) + (quat.i * quat.j));
+        let yaw_cos_pitch_cos = 1.0 - 2.0 * ((quat.j * quat.j) + (quat.k * quat.k));
+        let yaw = yaw_sin_pitch_cos.atan2(yaw_cos_pitch_cos);
+
+        // calculate pitch and roll
+        let pitch_sin = 2.0 * ((quat.real * quat.j) - (quat.k * quat.i));
+        let pitch: f64;
+        let roll: f64;
+        // for a pitch that is π / 2, we experience gimbal lock
+        // and must calculate roll based on the real rotation and yaw
+        if pitch_sin.abs() >= 1.0 {
+            pitch = (std::f64::consts::PI / 2.0).copysign(pitch_sin);
+            roll = (2.0 * quat.i.atan2(quat.real)) + yaw.copysign(pitch_sin);
```

Comment:
I admit I haven't checked the math myself, but [Wikipedia's reference implementation](https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles#Quaternion_to_Euler_angles_conversion) suggests that the roll is calculated the same way regardless of whether we've got gimbal lock. Are you sure about this one?

Reply:
Gautham set me straight: the Wikipedia page is wrong/misleading! :flushed: The next paragraph down from the code I linked is about "singularities" where that implementation breaks down. Gautham has instead showed my [this page](http://marc-b-reynolds.github.io/math/2017/04/18/TaitEuler.html), which is where he actually got his implementation from.